### PR TITLE
Add time derivative of a variable gradient

### LINKED ISSUE #10861: Add time derivative of gradient
## Rationale

Some physical quantities require the time derivative of gradient (for example the [displacement current](https://en.wikipedia.org/wiki/Displacement_current#Explanation) depends on the time derivative of the electric field, which in the Zapdos app the electric field is the gradient of the voltage)

## Description

We don't currently have _gradient_u_dot like we have a _u_dot.

## Impact

Would add the ability to model additional physics for users. According to Alex Lindsay it would not be difficult to add.

### PR DESCRIPTION
**Pull request by lindsayad**
Also refactor some variable related things.

 - Adds calculation of the time derivative of a gradient to moose variables
   - This is tested by an input file which solves the Coulomb gauge formulation of
     Maxwell's equations which contains the time derivative of the gradient of a coupled
     standard MooseVariable
     - Consequently this test also demonstrates coupling of a standard MooseVariable in a
       VectorKernel. Previously we had only tested coupling of a VectorMooseVariable to a
       Kernel
  - While building the new test, it was discovered that coupling a MooseVariable into a
    VectorKernel (or VectorIntegratedBC) would seg-fault because of a bad indexing
    operation, attributable to the fact that we iterate through j's up to
    `_phi.size()`. In the case of this new coupling, the size of the VectorMooseVariable
    is greater than the size of the coupled MooseVariable
    - Consequently, the off diagonal computing routines for kernels and boundary
      conditions were modified to actually take the moose variable as the method
      argument. Then j's can be iterated through up to the size of the coupled variable
  - Templated MooseVariableField was also split into .h and .C files to enable faster
    re-compilation when its methods are modified
    - Explicit instantiations of MooseVariableField<Real> (MooseVariable)
      and MooseVariableField<RealVectorValue> (VectorMooseVariable) are then provided at
      the end of the .C file
      - These explicit instantiations required generalization of the curl methods in moose
        variable and Assembly

Closes #10861

### CODE REVIEW

**Review by friedmud — CHANGES_REQUESTED** [review]
I'm against the API change to `offDiagJacobian`... it puts too much in the user's hands.

Let's brainstorm ideas about what can be done.

File: test/src/kernels/VectorCoeffReaction.C [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
```
Comment (andrsd) [inline test/src/kernels/VectorCoeffReaction.C]: old header

File: framework/src/base/MooseVariableField.C [hunk 2]
```diff
@@ -0,0 +1,752 @@
+#include "MooseVariableField.h"
```
Comment (andrsd) [inline framework/src/base/MooseVariableField.C]: missing header

File: test/include/bcs/VectorPenaltyDirichletBC.h [hunk 3]
```diff
@@ -0,0 +1,26 @@
+#ifndef VECTORPENALTYDIRICHLETBC_H
```
Comment (andrsd) [inline test/include/bcs/VectorPenaltyDirichletBC.h]: No header

File: test/src/bcs/VectorPenaltyDirichletBC.C [hunk 4]
```diff
@@ -0,0 +1,41 @@
+#include "VectorPenaltyDirichletBC.h"
```
Comment (andrsd) [inline test/src/bcs/VectorPenaltyDirichletBC.C]: No header

**Review by andrsd — DISMISSED** [review]

**Comment by lindsayad** [comment]
@friedmud We could add an additional argument to `computeOffDiagJacobian` corresponding to `_phi.size()`?

**Comment by friedmud** [comment]
@lindsayad that's kind of what I was thinking.  Can we make it default to something useful so we don't have to change every object in the whole system?

**Comment by friedmud** [comment]
BTW: Keep in mind that whatever solution we come up with here - might help us solve our `ArrayKernel` issues...

**Comment by lindsayad** [comment]
@friedmud What do you mean by changing every object in the system? You mean every object that overrides `computeOffDiagJacobian`? If we add a second argument, even with a default, we'll still have to change the method signatures of all of those objects that are overriding `computeOffDiagJacobian`.

For me a default argument doesn't make sense because all of these objects with `computeOffDiagJacobian` are getting called from `ComputeFullJacobianThread` where I would be providing the second argument.

**Comment by lindsayad** [comment]
@friedmud I think this should be ready to go

**Comment by friedmud** [comment]
Wait - I still don't understand why we have to change the interface here.  `_phi` should get traded out to be the coupled variable's `phi` when doing off-diagonal calculations... so the size of `_phi` should be the right thing.

**Comment by lindsayad** [comment]
Not if the coupled variable is a fundamentally different kind of variable, e.g. a `MooseVariable` vs. a `VectorMooseVariable`. In this case, `_phi` is of type `VariablePhiValue`; we can't covert it to type `VectorVariablePhiValue`

**Comment by friedmud** [comment]
Hmmm.  That makes sense - but is also a bummer.  I guess there is no confusion because if you're coupling to a vector variable then you're definitely going to need to use a different shape function for the computeQpOffDiagJacobian statements.

How well documented is this?  Is there documentation that spells out when you need to use one of the `_V_` member variables?  (BTW: I'm not in love with the `_V_` prefix... but I also can't think of much better besides writing out `_vector_` instead...)

**Comment by lindsayad** [comment]
Hmm...I'm not in love with `_V` because it breaks the C++ standard, so I'm going to change that. I think I will add some documentation to kernels/boundary conditions specifying when additional "phi" members ar needed

**Comment by lindsayad** [comment]
@friedmud 72fc25f has significant documentation added. The finite element principles come from the wiki

**Review by friedmud — APPROVED** [review]

**Comment by friedmud** [comment]
Looks good - thanks for all the doco!

**Pull request merged**